File: core/src/main/java/com/bindglam/utility/database/MySQLDatabase.java
package com.bindglam.utility.database;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import org.bukkit.configuration.ConfigurationSection;
import org.jetbrains.annotations.NotNull;

import java.sql.Connection;
import java.sql.SQLException;

public class MySQLDatabase implements Database {
    private HikariDataSource hikariDS;

    @Override
    public void connect(ConfigurationSection config) {
        HikariConfig hikariConfig = new HikariConfig();
        hikariConfig.setDriverClassName("com.mysql.cj.jdbc.Driver");
        hikariConfig.setJdbcUrl("jdbc:mysql://" + config.getString("host") + ":" + config.getString("port") + "/" + config.getString("database") + "?useUnicode=true&characterEncoding=utf8");
        hikariConfig.setUsername(config.getString("user"));
        hikariConfig.setPassword(config.getString("password"));
        hikariConfig.setConnectionTimeout(30000);
        hikariConfig.setIdleTimeout(10000);
        hikariConfig.setMinimumIdle(50); // 100
        hikariConfig.setMaximumPoolSize(1000); // 1000

        hikariDS = new HikariDataSource(hikariConfig);
    }

    @Override
    public void close() {
        if(hikariDS == null)
            return;

        if(!hikariDS.isClosed())
            hikariDS.close();
    }

    @Override
    public @NotNull Connection getConnection() {
        try {
            return hikariDS.getConnection();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void evictConnection(@NotNull Connection connection) {
        hikariDS.evictConnection(connection);
    }
}

File: core/src/main/kotlin/com/bindglam/utility/database/MySQLDatabase.kt
package com.bindglam.utility.database

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import org.bukkit.configuration.ConfigurationSection
import java.sql.Connection
import java.sql.SQLException

class MySQLDatabase : SQLDatabase {
    private var hikariDS: HikariDataSource? = null

    override fun connect(config: ConfigurationSection?) {
        val hikariConfig = HikariConfig()
        hikariConfig.driverClassName = "com.mysql.cj.jdbc.Driver"
        hikariConfig.jdbcUrl = "jdbc:mysql://" + config!!.getString("host") + ":" + config.getString("port") + "/" + config.getString("database") + "?useUnicode=true&characterEncoding=utf8"
        hikariConfig.username = config.getString("user")
        hikariConfig.password = config.getString("password")
        hikariConfig.connectionTimeout = 30000
        hikariConfig.idleTimeout = 10000
        hikariConfig.minimumIdle = 50 // 100
        hikariConfig.maximumPoolSize = 1000 // 1000

        hikariDS = HikariDataSource(hikariConfig)
    }

    override fun close() {
        if (hikariDS == null) return

        if (!hikariDS!!.isClosed) hikariDS!!.close()
    }

    override fun getConnection(): Connection {
        try {
            return hikariDS!!.connection
        } catch (e: SQLException) {
            throw RuntimeException(e)
        }
    }

    override fun evictConnection(connection: Connection) {
        hikariDS!!.evictConnection(connection)
    }
}
